### ISSUE TITLE
GCE heartbeat/fabric issues

### ISSUE THREAD

**Issue by cergfix** [aerospike/aerospike-server#97]
Hello,

We're experiencing some heartbeat  issues in Google Cloud (Aerospike Community edition). We're also getting quite a lot of cluster integrity failures that may be the result of this.

Here is some log info:
```
Oct 09 2015 08:54:41 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280067642 last 9184887394
Oct 09 2015 08:54:41 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280067792 last 9184887394
Oct 09 2015 08:54:41 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280067943 last 9184887394
Oct 09 2015 08:54:41 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280068093 last 9184887394
Oct 09 2015 08:54:41 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280068243 last 9184887394
Oct 09 2015 08:54:42 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280068393 last 9184887394
Oct 09 2015 08:54:42 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280068543 last 9184887394
Oct 09 2015 08:54:42 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280068693 last 9184887394
Oct 09 2015 08:54:42 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280068843 last 9184887394
Oct 09 2015 08:54:42 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280068994 last 9184887394
Oct 09 2015 08:54:42 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280069144 last 9184887394
Oct 09 2015 08:54:43 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280069294 last 9184887394
Oct 09 2015 08:54:43 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280069444 last 9184887394
Oct 09 2015 08:54:43 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280069594 last 9184887394
Oct 09 2015 08:54:43 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280069744 last 9184887394
Oct 09 2015 08:54:43 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280069895 last 9184887394
Oct 09 2015 08:54:43 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280070045 last 9184887394
Oct 09 2015 08:54:44 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280070460 last 9184887394
Oct 09 2015 08:54:44 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280070610 last 9184887394
Oct 09 2015 08:54:44 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280070760 last 9184887394
Oct 09 2015 08:54:44 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280070910 last 9184887394
Oct 09 2015 08:54:44 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280071060 last 9184887394
Oct 09 2015 08:54:44 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9280071210 last 9184887394
```

This all looks suspicious, but not fatal:
```
Oct 09 2015 09:10:56 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9281043077 last 9184887394
Oct 09 2015 09:10:56 GMT: INFO (hb): (hb.c::2224) hb expires but fabric says ALIVE: lasttime 40 node bb960d5f00a0142
Oct 09 2015 09:10:56 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9281043227 last 9184887394
Oct 09 2015 09:10:56 GMT: INFO (hb): (hb.c::2224) hb expires but fabric says ALIVE: lasttime 0 node bb960d5f00a0142
Oct 09 2015 09:10:57 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9281043377 last 9184887394
Oct 09 2015 09:10:57 GMT: INFO (hb): (hb.c::2224) hb expires but fabric says ALIVE: lasttime 40 node bb960d5f00a0142
Oct 09 2015 09:10:57 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9281043528 last 9184887394
Oct 09 2015 09:10:57 GMT: INFO (hb): (hb.c::2224) hb expires but fabric says ALIVE: lasttime 80 node bb960d5f00a0142
Oct 09 2015 09:10:57 GMT: INFO (hb): (hb.c::2202) hb considers expiring: now 9281043678 last 9184887394
```

until we get something like ```hb expires but fabric says DEAD```

This is happening mostly only between two nodes in the cluster. The gap is increasing. Looks like the first timestamp (uptime time) is correct while the second one is not (or not updated). 

How is this second timestamp set? Could it be dead socket that cannot recover?

**Comment by cergfix** [comment]
Ends up with 
```
Oct 10 2015 08:49:16 GMT: INFO (hb): (hb.c::2189) hb considers expiring: now 7332062980 last 7284498216
Oct 10 2015 08:49:16 GMT: DEBUG (fabric): (fabric.c:as_fabric_get_node_lasttime:1643) asking about node bb977ccf00a0142 good read 40 good write 40
Oct 10 2015 08:49:16 GMT: INFO (hb): (hb.c::2211) hb expires but fabric says ALIVE: lasttime 40 node bb977ccf00a0142
Oct 10 2015 08:49:16 GMT: INFO (hb): (hb.c::2189) hb considers expiring: now 7332062980 last 7308255382
Oct 10 2015 08:49:16 GMT: DEBUG (fabric): (fabric.c:as_fabric_get_node_lasttime:1643) asking about node bb9a4d9f00a0142 good read 80 good write 80
Oct 10 2015 08:49:16 GMT: INFO (hb): (hb.c::2211) hb expires but fabric says ALIVE: lasttime 80 node bb9a4d9f00a0142
Oct 10 2015 08:49:16 GMT: DEBUG (fabric): (fabric.c:fabric_worker_add:1328) worker_fabric_add: adding fd 130 to worker id 8 notefd 31
Oct 10 2015 08:49:16 GMT: DEBUG (fabric): (fabric.c:fabric_worker_fn:1476) epoll : error, will close: fb 0x7f2321447008 fd 130 errno 2
Oct 10 2015 08:49:16 GMT: DEBUG (fabric): (fabric.c:fabric_buffer_release:537) removing fb 0x7f2321447008 connected 1 from fne 0x7f230d00eb58 connected_fb_hash
Oct 10 2015 08:49:16 GMT: DEBUG (fabric): (fabric.c:fabric_buffer_release:541) removed fb 0x7f2321447008 from connected_fb_hash
Oct 10 2015 08:49:16 GMT: DEBUG (fabric): (fabric.c:fabric_buffer_release:571) dropping message: TCP connection close with writable bytes 306
Oct 10 2015 08:49:16 GMT: INFO (hb): (hb.c::2189) hb considers expiring: now 7332063130 last 7284498216
Oct 10 2015 08:49:16 GMT: DEBUG (fabric): (fabric.c:as_fabric_get_node_lasttime:1643) asking about node bb977ccf00a0142 good read 40 good write 40
Oct 10 2015 08:49:16 GMT: INFO (hb): (hb.c::2211) hb expires but fabric says ALIVE: lasttime 40 node bb977ccf00a0142
Oct 10 2015 08:49:16 GMT: INFO (hb): (hb.c::2189) hb considers expiring: now 7332063130 last 7308255382
Oct 10 2015 08:49:16 GMT: DEBUG (fabric): (fabric.c:as_fabric_get_node_lasttime:1643) asking about node bb9a4d9f00a0142 good read 40 good write 40
Oct 10 2015 08:49:16 GMT: INFO (hb): (hb.c::2211) hb expires but fabric says ALIVE: lasttime 40 node bb9a4d9f00a0142
Oct 10 2015 08:49:16 GMT: INFO (hb): (hb.c::2319) HB node bb9a4d9f00a0142 in different cluster - succession lists don't match
Oct 10 2015 08:49:17 GMT: INFO (hb): (hb.c::2189) hb considers expiring: now 7332063280 last 7332057973
Oct 10 2015 08:49:17 GMT: DEBUG (fabric): (fabric.c:as_fabric_get_node_lasttime:1643) asking about node bb9dbd2f00a0142 good read 5371 good write 3408
Oct 10 2015 08:49:17 GMT: INFO (hb): (hb.c::2202) hb expires but fabric says DEAD: node bb9dbd2f00a0142
Oct 10 2015 08:49:17 GMT: DEBUG (hb): (hb.c:as_hb_monitor_reduce:2227) node has actually expired according to heartbeat node bb9dbd2f00a0142
Oct 10 2015 08:49:17 GMT: INFO (hb): (hb.c::2189) hb considers expiring: now 7332063280 last 7284498216
Oct 10 2015 08:49:17 GMT: DEBUG (fabric): (fabric.c:as_fabric_get_node_lasttime:1643) asking about node bb977ccf00a0142 good read 0 good write 0
Oct 10 2015 08:49:17 GMT: INFO (hb): (hb.c::2211) hb expires but fabric says ALIVE: lasttime 0 node bb977ccf00a0142
Oct 10 2015 08:49:17 GMT: INFO (hb): (hb.c::2189) hb considers expiring: now 7332063280 last 7308255382
Oct 10 2015 08:49:17 GMT: DEBUG (fabric): (fabric.c:as_fabric_get_node_lasttime:1643) asking about node bb9a4d9f00a0142 good read 0 good write 0
Oct 10 2015 08:49:17 GMT: INFO (hb): (hb.c::2211) hb expires but fabric says ALIVE: lasttime 0 node bb9a4d9f00a0142
Oct 10 2015 08:49:17 GMT: DEBUG (hb): (hb.c:as_hb_monitor_thr:2363) internal snub: node bb9dbd2f00a0142 waits 52500
Oct 10 2015 08:49:17 GMT: DEBUG (hb): (hb.c:as_hb_snub:958) snub : node bb9dbd2f00a0142 time 52500
Oct 10 2015 08:49:17 GMT: INFO (hb): (hb.c::2382) removing node on heartbeat failure: bb9dbd2f00a0142
Oct 10 2015 08:49:17 GMT: INFO (fabric): (fabric.c::1786) fabric: node bb9dbd2f00a0142 departed
Oct 10 2015 08:49:17 GMT: INFO (fabric): (fabric.c::1696) fabric disconnecting node: bb9dbd2f00a0142
Oct 10 2015 08:49:17 GMT: DEBUG (fabric): (fabric.c:fabric_node_disconnect:1720) fabric_node_disconnect(0xbb9dbd2f00a0142): fne: 0x7f230d00eb58 : xmit buffer queue empty
Oct 10 2015 08:49:17 GMT: DEBUG (fabric): (fabric.c:fabric_node_disconnect:1729) fabric: dropping message to now-gone (heartbeat fail) node bb9dbd2f00a0142
Oct 10 2015 08:49:17 GMT: DEBUG (fabric): (fabric.c:fabric_node_disconnect:1729) fabric: dropping message to now-gone (heartbeat fail) node bb9dbd2f00a0142
Oct 10 2015 08:49:17 GMT: DEBUG (fabric): (fabric.c:fabric_node_disconnect:1729) fabric: dropping message to now-gone (heartbeat fail) node bb9dbd2f00a0142
Oct 10 2015 08:49:17 GMT: DEBUG (fabric): (fabric.c:fabric_node_disconnect:1729) fabric: dropping message to now-gone (heartbeat fail) node bb9dbd2f00a0142
Oct 10 2015 08:49:17 GMT: DEBUG (fabric): (fabric.c:fabric_node_disconnect:1729) fabric: dropping message to now-gone (heartbeat fail) node bb9dbd2f00a0142
Oct 10 2015 08:49:17 GMT: DEBUG (fabric): (fabric.c:fabric_node_disconnect:1729) fabric: dropping message to now-gone (heartbeat fail) node bb9dbd2f00a0142
Oct 10 2015 08:49:17 GMT: DEBUG (fabric): (fabric.c:fabric_node_disconnect:1729) fabric: dropping message to now-gone (heartbeat fail) node bb9dbd2f00a0142
```

**Comment by wchu-citrusleaf** [comment]
The logs indicate the nodes in the cluster no longer can communicate with each other.

From other experience we have seen in GCE environment, the instance live-migration causes network brownouts, thus splitting the cluster. This is particularly true for memory intensive systems, where all memory-in-use needs to migrate out.

https://cloud.google.com/compute/docs/zones?hl=en#maintenance

One operational alternative is to increase the heartbeat interval so the cluster is more lenient to brown-out.

Reaching out to GCE support  also advised.

**Comment by ivansmf** [comment]
I work on Google Cloud performance. We have worked with Aerospike to make them more resilient to clock skew, which is the actual root cause for the problem you saw. They have a fix that should reach their stable branch soon (I have no ETA). We deployed a patched system to our clusters and run a stress workload for 3 days, migrating all VMs every 2h and it held well, despite some throughput fluctuation.

Until the fix is available one mitigation is to use smaller instances. If you build your cluster using 4 core instances this problem should go away. Smaller instances have less memory, which causes migrations to become shorter, which reduces the period in which we have clock skew, making much less likely that problems like this hit you.

I hope this helps.

**Comment by sunilvirus** [comment]
Hi, I will send you a dev build with the enhancements.
I will use the email that is mentioned in your profile.
We will be releasing an official build shortly.

**Comment by sunilvirus** [comment]
The new official build 3.7.0 which is about to come out will have all the fixes.

**Comment by bbulkow** [comment]
These fixes are targeted at being optional in the 3.7 build and controlled by a config option - as we are still studying the effect in production environments - so be sure to check the release notes carefully.

**Comment by wchu-citrusleaf** [comment]
now in 3.7.0.1  and above.
[AER-4500] - (Clustering) New cluster recovery policy paxos-recovery-policy=auto-reset-master.
